accord-core/src/main/java/accord/primitives/Sortables.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    @Inline
+    static <Input extends SortableKey, T> T foldl(AbstractKeys<Input, ?> inputs, AbstractRanges<?> matching, IndexedFold<? super Input, T> fold, T initialValue)
+    {
+        return Helper.foldl(AbstractKeys::findNextIntersection, Helper::findLimit, inputs, matching, fold, initialValue);
+    }
+
+    @Inline
+    static <Input extends SortableKey> long foldl(AbstractKeys<Input, ?> inputs, AbstractRanges<?> matching, IndexedFoldToLong<? super Input> fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.foldl(AbstractKeys::findNextIntersection, Helper::findLimit, inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends Sortable> long foldl(Sortables<Input, ?> inputs, AbstractRanges<?> matching, IndexedFoldToLong<? super Input> fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.foldl(Sortables::findNextIntersection, Helper::findLimit, inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends Sortable> long foldl(Sortables<Input, ?> inputs, AbstractKeys<?, ?> matching, IndexedFoldToLong<? super Input> fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.foldl(Sortables::findNextIntersection, (ls, li, rs, ri) -> li + 1,
+                inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends RoutingKey, Matching extends Sortable> long foldl(AbstractKeys<Input, ?> inputs, Sortables<Matching, ?> matching, IndexedFoldToLong<? super Input> fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.foldl((ls, li, rs, ri) -> SortedArrays.swapHighLow32b(rs.findNextIntersection(ri, ls, li)), (ls, li, rs, ri) -> li + 1,
+                inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends Sortable> long rangeFoldl(Sortables<Input, ?> inputs, AbstractRanges<?> matching, IndexedRangeFoldToLong fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.rangeFoldl(Sortables::findNextIntersection, (ls, li, rs, ri) -> li + 1,
+                inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends Sortable> long rangeFoldl(Sortables<Input, ?> inputs, AbstractKeys<?, ?> matching, IndexedRangeFoldToLong fold, long param, long initialValue, long terminalValue)
+    {
+        return Helper.rangeFoldl(Sortables::findNextIntersection, (ls, li, rs, ri) -> li + 1,
+                inputs, matching, fold, param, initialValue, terminalValue);
+    }
+
+    @Inline
+    static <Input extends Sortable> long foldlMissing(Sortables<Input, ?> inputs, Sortables<Input, ?> notMatching, IndexedFoldToLong<? super Input> fold, long param, long initialValue, long terminalValue)
```

Comment:
This was cut/paste from elsewhere, with some minor edits to generify it. I agree it could do with some randomised tests, but would prefer to focus elsewhere for now as time is rapidly getting away from us.